### ISSUE TITLE
No output image

### ISSUE THREAD

**Issue by bensleveritt** [dfrankland/nixos-rpi-sd-image#1]
First off. Great work! I've been stuck with this exact problem for a week or so, trying to make NixOS for various Raspberry Pis, but without making much headway. I'm not quite knowledgable enough about Nix (or cross-compilation) to have one clear/simple idea about what to do to make it work. Instead I've been bouncing about between complex solutions and weird VM installations. 

Your work seems most straight forward, so thanks!

When I run `docker-compose up` the image builds and completes successfully, but there's no image in the output folder.

Is there anything I do to figure out why? Is it possible there's a permissions issue, and the image is still inside the Docker image?

**Comment by bensleveritt** [comment]
The final output is this:

```
nixos-rpi-sd-image_1  | 31457280 bytes (31 MB, 30 MiB) copied, 0.214489 s, 147 MB/s
nixos-rpi-sd-image_1  | /nix/store/jg5qbcmnyadl487ndwhpwvhm1ncn9iqy-nixos-sd-image-21.03pre260194.ea3638a3fb2-aarch64-linux.img/nix-support/system
nixos-rpi-sd-image_1  | Found img file: /nix/store/jg5qbcmnyadl487ndwhpwvhm1ncn9iqy-nixos-sd-image-21.03pre260194.ea3638a3fb2-aarch64-linux.img/nix-support/system
nixos-rpi-sd-image_nixos-rpi-sd-image_1 exited with code 0
```

It copies a `system` text file into the output rather than the image.

**Comment by dfrankland** [comment]
Oh! That is interesting. I wonder why `/nix/store/jg5qbcmnyadl487ndwhpwvhm1ncn9iqy-nixos-sd-image-21.03pre260194.ea3638a3fb2-aarch64-linux.img/nix-support/system` is getting printed out, I haven't seen that before. What is the contents of the `system` text file?

If you want to investigate, I recommend that you build the Docker image, then run it interactively with bash. Then you can run the script to generate an image and see if you can find the `.img` file. Once we figure out what's going on we can automate that in the script.

**Comment by bensleveritt** [comment]
Thanks for your response!

So I built the container with the `generate-sd-image.sh` line commented out from the Dockerfile, and ran it with `docker run -it <result of docker build nixos-rpi-sd-image>` and the error I get is:

```
...
building '/nix/store/1q0a2zkk8y4dxz84js71qfdx10x9aqsr-ext4-fs.img.zst.drv'...
Preparing store paths for image...
Creating an EXT4 image of 3305017344 bytes (numInodes=101744, numDataBlocks=603401)
semop(1): encountered an error: Function not implemented
builder for '/nix/store/1q0a2zkk8y4dxz84js71qfdx10x9aqsr-ext4-fs.img.zst.drv' failed with exit code 1
cannot build derivation '/nix/store/mirbr8g4pg6r1cymyxwc199avhl32nqp-nixos-sd-image-21.03pre260194.ea3638a3fb2-aarch64-linux.img.drv': 1 dependencies couldn't be built
error: build of '/nix/store/mirbr8g4pg6r1cymyxwc199avhl32nqp-nixos-sd-image-21.03pre260194.ea3638a3fb2-aarch64-linux.img.drv' failed
```

This seems to be [a similar error to this](https://linux.debian.maint.glibc.narkive.com/UxqeBXag/bug-965109-libc6-semop-1-encountered-an-error-function-not-implemented-in-fakeroot-under-qemu). Seems to be something that's fixed in upstream, but that was 6 months ago (not determined if the fix is in yet, and if it is, does Nixpkgs have it?)

And that's about as far as I've got. As we're handing off building to QEMU in a Docker container, wouldn't it be possible to make a Nix expression that could do the same, thus skipping Docker entirely (I've been trying a number of different ways, but not managed an image yet)?

Anyway, thanks for looking at this!

**Comment by bensleveritt** [comment]
Just to add: I switched to unstable and updated nixpkgs.

```
RUN nix-channel --add https://nixos.org/channels/nixpkgs-unstable nixpkgs
RUN nix-channel --update
```

It errored out the same way. I'll try an older version...

**Comment by bensleveritt** [comment]
Well that's damn weird, it's working now (yay!)

I suspect it's from running the binfmt registration again (`docker run --rm --privileged multiarch/qemu-user-static --reset -p yes`), as I've only really been looking at the variations of multiarch today. Not sure, and I'm not sure I can reproduce to try again.

Thanks again for your work, you have the most straight forward approach so far! Now I just need to work out how to make images for Rpi0/2s.

**Comment by dfrankland** [comment]
I figured out the `system` file problem. Looks like that only happens on MacOS. I added that case to the `generate-sd-image.sh` script now:
https://github.com/dfrankland/nixos-rpi-sd-image/commit/f0ad101a1d555e2310d4e238d55b7f068e2f6951

Yep, I've encountered that same problem with `semop` and every time I do, rerunning the Docker command for binfmt registration fixes it. I have a feeling it's necessary to do every time one's machine is restarted 🤔

Glad you were finally able to get something working. It can be frustrating to not make progress, especially after a week!
